### ISSUE TITLE
Autoeviction for Task Inputs

### ISSUE THREAD

**Issue by WardLT** [exalearn/colmena#136]
We don't auto-evict data from proxystore for inputs from completed tasks, which can lead to memory/disk use issues. 

We could...
- Make autoeviction on read for the inputs from the compute worker, which will break task restarting
- Make autoeviction after the tasks complete on the worker
- ...

**Comment by gpauloski** [comment]
Some other ideas:
* Add a callback to `Result.__del__` which evicts all the associated proxies. Not my favourite idea because one could pickle a `Result` and send it to another process causing the value to be GC collected on the original process.
* Add a callback to the `Future` of the task result.
  * Would address the task restarting problem.
  * The decision to evict the inputs could be tied to the `keep_inputs` flag. This would mean that `inputs` eviction is done on the task server, and `value` eviction would be done when access by the Thinker. 
  * We would need to make it clear in some way that if `keep_inputs=True` then managing input proxies is up to the user (and potentially add a convenience callback like `Result.cleanup()` or something to clean up those associated proxies).

**Comment by WardLT** [comment]
Do you think we can evict from the worker process? E.g., have it evict right before exiting a successful task? 

https://github.com/exalearn/colmena/blob/master/colmena/models/methods.py#L99

**Comment by gpauloski** [comment]
I can see why it would be beneficial to do that on the worker. The tricky part is what happens if two tasks A and B get the same proxy and A completes execution be B starts? Then A may evict the data before B gets it.

We can probably guarantee that if its Colmena auto-proxied objects but user-proxied objects would be more challenging to reference count.

**Comment by WardLT** [comment]
Great point. Maybe we can set autoevict if auto-proxying but don't otherwise 🤔 

That's what I've got on an uncommitted fork of Colmena now and it seems OK. 

Autoevict does break Parsl's task restarting, but I haven't needed that too.

**Comment by gpauloski** [comment]
Yeah, Parsl task restarting would be the only thing I'd be worried about there. As long as users can opt out if needed I think it should be okay.
